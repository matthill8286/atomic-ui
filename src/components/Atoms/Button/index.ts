export * from './Button'
export * from './Button.interface'
export * from './IconButton'
