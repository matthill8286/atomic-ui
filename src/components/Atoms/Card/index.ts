export * from './CardV2'
export * from './Card.interface'
export { StyledCard } from './Card.styled'
