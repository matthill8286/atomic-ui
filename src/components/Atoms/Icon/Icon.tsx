import { FlexAlignItemsOptions } from '@/components/Helper/FlexBox'
import { padding } from '@/styles/sc-shared-functions'
import { css, styled } from '@/styles/styled'
import { PaddingProps, Size, ThemeColors } from '@/types'

export type RotationValues = 0 | 45 | 90 | 180 | 270 | -45 | -90 | -180 | -270

export interface IconProps {
  rotate?: RotationValues
  color?: ThemeColors
  withBorder?: boolean
  padding?: PaddingProps['padding']

  width?: number | keyof Omit<Size, 'xxxs'>
  height?: number | keyof Omit<Size, 'xxxs'>
  animate?: boolean
  display?: string
  alignSelf?: FlexAlignItemsOptions
}

const DefaultSize = '25px'

export const Icon = styled.div<IconProps>(
  ({
    theme,
    color,
    width,
    height,
    rotate = 0,
    animate = false,
    display = 'flex',
    alignSelf,
    withBorder,
  }) => css`
    ${display && `display: ${display};`};
    ${padding};
    ${alignSelf && `align-self: ${alignSelf};`};
    ${withBorder &&
      `border: 2px solid ${theme.color.grey2};
        border-radius: 50%;
        padding: ${theme.spacing.base.xs};`}
    svg {
      width: ${width ? getSizeInPx(width, theme.spacing.base) : DefaultSize};
      height: ${height ? getSizeInPx(height, theme.spacing.base) : DefaultSize};
      transform: rotate(${rotate}deg);
      fill: ${color ? theme.color[color] : 'currentColor'};
      ${animate &&
        css`
          transition: transform 0.3s ease-in-out;
        `};
    }
    svg g,
    svg p {
      fill: ${color ? theme.color[color] : 'currentColor'};
    }
  `
)

const getSizeInPx = (
  size: number | keyof Omit<Size, 'xxxs'>,
  themeSizes: Omit<Size, 'xxxs'>
): string => {
  if (typeof size === 'number') {
    return `${size}px`
  } else {
    return themeSizes[size] || `${DefaultSize}px`
  }
}
