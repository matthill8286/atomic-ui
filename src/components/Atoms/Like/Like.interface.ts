import { ComponentType, ReactElement } from 'react'
import { MouseEvent } from 'react'
import { ThemeColors, ThemeFontLineHeight, ThemeFontSizes } from '@/types/theme'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type LikeAs = string | JSX.IntrinsicElements | ComponentType<any>
export type LikeSize = 'small' | 'large'
type CursorType = 'initial' | 'pointer'

export interface LikeProps {
  bold?: boolean
  className?: string
  cursor?: CursorType
  color?: ThemeColors
  decorationColor?: ThemeColors
  disabled?: boolean
  fontSize?: ThemeFontSizes
  iconLeft?: ReactElement
  iconRight?: ReactElement
  inline?: boolean
  lineHeight?: ThemeFontLineHeight
  onClick?: (e: MouseEvent<HTMLElement>) => void
  scale?: LikeSize
  copy?: string
  likes?: string
  underline?: boolean
}

export interface StyledLikeProps {
  color?: ThemeColors
  cursor?: CursorType
  hasIconLeft: boolean
  hasIconRight: boolean
  isBold: boolean
  isDisabled: boolean
  isInline: boolean
  isLarge: boolean
  isSmall: boolean
  isUnderlined?: boolean
  copy?: string
  likes?: string
}
