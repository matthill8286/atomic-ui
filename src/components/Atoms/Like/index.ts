export * from './Like'
export * from './Like.interface'
