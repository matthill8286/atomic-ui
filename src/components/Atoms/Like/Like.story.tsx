import { boolean, select, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import * as React from 'react'
import { Icon } from '@/components/Atoms/Icon'
import { CopyText } from '@/components/Atoms/Typography'
import { saiyanTheme } from '@/styles/sc-vars-saiyan'
import { OtherUseful } from '@matthill8286/atomic-icon-library'
import { ThemeColors, ThemeFontSizes } from '@/types/theme'
import { Like } from './Like'
import { LikeProps, LikeSize } from './Like.interface'

const scales = ['large', 'small']
const sizes: ThemeFontSizes[] = ['xxxs', 'xxs', 'xs', 'sm', 'md', 'lg', 'xl', 'xxl', 'xxxl']

const knobs = (): Partial<LikeProps> => {
  return {
    color: select('Color', Object.keys(saiyanTheme.color), 'grey4') as ThemeColors,
    disabled: boolean('Set disabled', false),
    bold: boolean('Bold', true),
  }
}

storiesOf('Design System/Atoms/Like', module)
  .add('Like', () => {
    const scale = select('Type scale', scales, 'large') as LikeSize
    const sampleCopy = text('Example copy', `found this useful`)
    const sampleLikes = text('Example likes', `38 `)
    const underline = boolean('Underline', true)

    return (
      <Like {...knobs()} scale={scale} likes={sampleLikes} underline={underline}>
        {sampleCopy}
      </Like>
    )
  })
  .add('with Icon', () => {
    const scale = select('Type scale', scales, 'large') as LikeSize
    const sampleCopy = text('Example copy', `found this useful`)
    const sampleLikes = text('Example likes', `55`)
    const iconPositions = ['left', 'right', 'both']
    const iconPosition = select('Icon Position', iconPositions, 'left')
    const iconColor = select('Icon Color', Object.keys(saiyanTheme.color), 'grey4') as ThemeColors
    const underline = boolean('Underline', true)

    return (
      <Like
        {...knobs()}
        scale={scale}
        underline={underline}
        likes={sampleLikes}
        {...((iconPosition === 'left' || iconPosition === 'both') && {
          iconLeft: (
            <Icon color={iconColor}>
              <OtherUseful />
            </Icon>
          ),
        })}
        {...((iconPosition === 'right' || iconPosition === 'both') && {
          iconRight: (
            <Icon color={iconColor}>
              <OtherUseful />
            </Icon>
          ),
        })}>
        {sampleCopy}
      </Like>
    )
  })
  .add('inline Like no Icon', () => {
    const sampleCopy = text('Example copy', `Lorem ipsum Like`)
    const fontSize = select('Size', sizes, sizes[3])
    const underline = boolean('Underline', true)
    return (
      <CopyText fontSize={fontSize} tag="div">
        <Like {...knobs()} fontSize={fontSize} inline underline={underline}>
          {sampleCopy}
        </Like>
        {' extended copy'}
      </CopyText>
    )
  })
