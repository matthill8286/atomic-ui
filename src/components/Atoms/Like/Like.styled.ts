import { CopyText } from '@/components/Atoms/Typography'
import { styled } from '@/styles/styled'
import { Theme } from '@/types/theme'
import { StyledLikeProps } from './Like.interface'

export const StyledLike = styled.div<StyledLikeProps>`
  display: ${({ isInline }) => (isInline ? `inline-flex` : 'flex')};
  align-items: center;
  width: fit-content;
  color: ${({ theme, isDisabled, isInline }) =>
    (isDisabled && theme.color.grey2) || (isInline && 'inherit') || theme.color.grey4};
  opacity: ${({ theme, isDisabled }) => (isDisabled ? theme.polished.inactive : `1.0`)};
  text-decoration: none;
  cursor: ${({ isDisabled, cursor }) => (isDisabled ? 'initial' : cursor || 'pointer')};
  position: relative;
  padding: ${({ theme }) => `${theme.spacing.base.xxs} ${theme.spacing.base.xs}`};
  border-bottom: ${({ isUnderlined, isInline, theme }) =>
    isUnderlined && isInline ? `1px solid ${theme.color.grey2}` : ''};

  ${({ isDisabled, isInline }) =>
    !isDisabled &&
    !isInline &&
    `
    :hover {
      &::after {
        transform-origin: left center;
        transform: scale(1, 1);
      }
    }
  `}
`

/**
 * Get the width/height of the icon depending if the Like is small or large.
 *
 * @param theme
 * @param isSmall
 */
const getIconSize = (theme: Theme, isSmall: boolean) => {
  return isSmall ? theme.font.lineHeight.xxs : theme.font.lineHeight.xs
}

const getIconMargin = (theme: Theme) => {
  return theme.spacing.base.xxs
}

export const StyledIconWrapper = styled.span<StyledLikeProps>`
  margin-left: ${({ hasIconRight, theme }) => (hasIconRight ? getIconMargin(theme) : '0px')};
  margin-right: ${({ hasIconLeft, theme }) => (hasIconLeft ? getIconMargin(theme) : '0px')};

  & svg {
    height: ${({ theme, isSmall }) => getIconSize(theme, isSmall)};
    width: ${({ theme, isSmall }) => getIconSize(theme, isSmall)};
  }
`

export const Text = styled(CopyText)<{ isInline: boolean }>`
  display: ${props => (props.isInline ? 'inline' : 'block')};
`

export const Likes = styled(CopyText)<{ isInline: boolean }>`
  display: ${props => (props.isInline ? 'inline' : 'block')};
  margin-right: 5px;
`
