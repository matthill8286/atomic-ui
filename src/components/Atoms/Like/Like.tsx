import React from 'react'
import { LikeProps, StyledLikeProps } from './Like.interface'
import { Likes, StyledIconWrapper, StyledLike, Text } from './Like.styled'

export const Like: React.FC<LikeProps> = ({
  bold = false,
  children,
  className,
  cursor,
  color,
  decorationColor,
  disabled = false,
  fontSize,
  iconLeft,
  iconRight,
  inline = false,
  lineHeight,
  onClick,
  scale = 'small',
  copy,
  likes,
  underline = true,
  ...props
}): JSX.Element | null => {
  const isSmall = scale === 'small'
  const styleProps: StyledLikeProps = {
    color: decorationColor || color,
    cursor,
    hasIconLeft: !!iconLeft,
    hasIconRight: !!iconRight,
    isBold: bold,
    isDisabled: disabled,
    isInline: inline,
    isLarge: scale === 'large',
    isSmall,
    isUnderlined: underline,
  }

  if (!likes) {
    return null
  }

  return (
    <StyledLike className={className} onClick={onClick} {...styleProps} {...props}>
      {iconLeft && <StyledIconWrapper {...styleProps}>{iconLeft}</StyledIconWrapper>}
      <Likes
        tag="div"
        fontSize={fontSize || (isSmall ? 'xs' : 'sm')}
        lineHeight={lineHeight}
        bold={bold}
        color={color}
        isInline={inline}>
        {likes}
      </Likes>
      <Text
        tag="div"
        fontSize={fontSize || (isSmall ? 'xs' : 'sm')}
        lineHeight={lineHeight}
        color={color}
        isInline={inline}>
        {children}
      </Text>
      {iconRight && <StyledIconWrapper {...styleProps}>{iconRight}</StyledIconWrapper>}
    </StyledLike>
  )
}
