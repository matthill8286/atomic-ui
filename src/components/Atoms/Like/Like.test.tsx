import React from 'react'
import { Icon } from '@/components/Atoms/Icon'
import { IconArticle } from '@matthill8286/atomic-icon-library'
import { renderWithTheme } from '@/testRenderer'
import { Like } from './Like'

describe('Like', () => {
  it('renders correctly', () => {
    const tree = renderWithTheme(<Like scale="small">Test Like</Like>)
    expect(tree).toMatchSnapshot()
  })
  it('renders with left Icon', () => {
    const tree = renderWithTheme(
      <Like
        scale="large"
        iconLeft={
          <Icon>
            <IconArticle />
          </Icon>
        }>
        Test Like with Icon
      </Like>
    )
    expect(tree).toMatchSnapshot()
  })
  it('renders with right Icon', () => {
    const tree = renderWithTheme(
      <Like
        scale="small"
        iconRight={
          <Icon>
            <IconArticle />
          </Icon>
        }>
        Test Like with Icon
      </Like>
    )
    expect(tree).toMatchSnapshot()
  })
})
