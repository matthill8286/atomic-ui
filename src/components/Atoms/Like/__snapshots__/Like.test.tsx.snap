// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Like renders correctly 1`] = `null`;

exports[`Like renders with left Icon 1`] = `null`;

exports[`Like renders with right Icon 1`] = `null`;
